### DISCUSSION TITLE
Can't fix the cors error even after following the docs

### DISCUSSION THREAD

**Discussion by krankush008** [vercel/community#4337] [Help] [answered]
### Summary

Hi guys, I have tried every possible ways to fix the cors error but can't fix it. My server side looks like this 

import cors from "cors";
import express from "express";
import BeetingUsers from './model/user.js';

const app = express()
const port = 8384
app.use(cors(
  {
     origin: "https://luminous-jelly-76ea42.netlify.app",
     methods: ["POST", "GET"],
     credentials: true
  }
))
app.use(express.json());

app.options('*', (req, res) => {
  if (req.method === 'OPTIONS') {
    return res.status(200).json({ body: "OK" });
  }
});

app.listen(port, () => {
  console.log(`Example app listening on port ${port}`)
})

app.get("/", (req, res) => {
  res.json("hello");
})

app.post('/signup', async(req, res) => {
    const user = new BeetingUsers({
       username:JSON.parse(req.body.body).username,
       email:JSON.parse(req.body.body).email,
       password:JSON.parse(req.body.body).password
    })

    const alreadyRegisteredUser = await BeetingUsers.findOne({ email: JSON.parse(req.body.body).email });

    if (alreadyRegisteredUser) {
      return res.status(409).json({ message: 'This email is already registered. Kindly login through the login page' });
    }

    res.setHeader('Access-Control-Allow-Origin', 'https://luminous-jelly-76ea42.netlify.app');

    if(req.method === 'OPTIONS') {
      return res.status(200).json(({
          body: "OK"
      }))
  }

    await user.save()
    .then(result=>{
      res.status(200).json({
        message: 'user registered'
      })
    })
    .catch(err=>{
      res.status(500).json({
        error:err
      })
    })

})

app.get('/signin', async (req, res) => {
  try {
    const userEmail =  req.query.email; 
    const user = await BeetingUsers.findOne({ email: userEmail });
    if (!user) {
      return res.status(404).json({ message: 'kindly register and then try logging in' });
    }
    if(user.password!=req.query.password){
      return res.status(404).json({ message: 'wrong password' });
    } 

    res.setHeader('Access-Control-Allow-Origin', 'https://luminous-jelly-76ea42.netlify.app');

    if(req.method === 'OPTIONS') {
      return res.status(200).json(({
          body: "OK"
      }))
  }
    return res.status(200).json({ message: 'user found' });
  } catch (error) {
    return res.status(500).json({ message: 'Internal server error' });
  }

})

and my vercel.json looks like this

{
    "version": 2,
    "builds": [
        {"src": "./index.js", "use": "@vercel/node"}
    ],
    "routes": [
        {
          "src": "/(.*)",
          "dest": "./index.js"
        }
    ],
     "headers": [
          {
            "source": "/(.*)",
            "headers": [
              { "key": "Access-Control-Allow-Credentials", "value": "true" },
              { "key": "Access-Control-Allow-Origin", "value": "*" },
              { "key": "Access-Control-Allow-Methods", "value": "GET,OPTIONS,PATCH,DELETE,POST,PUT" },
              { "key": "Access-Control-Allow-Headers", "value": "X-CSRF-Token, X-Requested-With, Accept, Accept-Version, Content-Length, Content-MD5, Content-Type, Date, X-Api-Version" }
            ]
          }
        ]  
}

Kindly help me resolve this issue because am struggling to figure out the problem.

### Steps to Reproduce

This is my github link.

https://github.com/krankush008/bettingBackend/tree/main

**Comment by amyegan** [comment, marked as answer]
It looks like you combined legacy [`builds`](https://vercel.com/docs/projects/project-configuration#builds) and [`routes`](https://vercel.com/docs/projects/project-configuration#routes) settings with [`headers`](https://vercel.com/docs/projects/project-configuration#headers). That might be part of the problem. For Express apps, I recommend following this guide instead: https://vercel.com/guides/using-express-with-vercel

Alternatively, you might have better luck adding the headers in your Express code. This discussion has some good suggestions: https://stackoverflow.com/questions/34644622/how-to-manage-cors-policy-properly-in-express
